zcash_primitives/src/note_encryption.rs

```diff
@@ -0,0 +1,1079 @@
+
+const COMPACT_NOTE_SIZE: usize = (
+    1  + // version
+    11 + // diversifier
+    8  + // value
+    32
+    // rcv
+);
+const NOTE_PLAINTEXT_SIZE: usize = COMPACT_NOTE_SIZE + 512;
+const OUT_PLAINTEXT_SIZE: usize = (
+    32 + // pk_d
+    32
+    // esk
+);
+const ENC_CIPHERTEXT_SIZE: usize = NOTE_PLAINTEXT_SIZE + 16;
+const OUT_CIPHERTEXT_SIZE: usize = OUT_PLAINTEXT_SIZE + 16;
+
+/// Format a byte array as a colon-delimited hex string.
+///
+/// Source: https://github.com/tendermint/signatory
+/// License: MIT / Apache 2.0
+fn fmt_colon_delimited_hex<B>(f: &mut fmt::Formatter<'_>, bytes: B) -> fmt::Result
+where
+    B: AsRef<[u8]>,
+{
+    let len = bytes.as_ref().len();
+
+    for (i, byte) in bytes.as_ref().iter().enumerate() {
+        write!(f, "{:02x}", byte)?;
+
+        if i != len - 1 {
+            write!(f, ":")?;
+        }
+    }
+
+    Ok(())
+}
+
+/// An unencrypted memo received alongside a shielded note in a Zcash transaction.
+#[derive(Clone)]
+pub struct Memo([u8; 512]);
+
+impl fmt::Debug for Memo {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Memo(")?;
+        match self.to_utf8() {
+            Some(Ok(memo)) => write!(f, "{}", memo)?,
+            _ => fmt_colon_delimited_hex(f, &self.0[..])?,
+        }
+        write!(f, ")")
+    }
+}
+
+impl Default for Memo {
+    fn default() -> Self {
+        // Empty memo field indication per ZIP 302
+        let mut memo = [0u8; 512];
+        memo[0] = 0xF6;
+        Memo(memo)
+    }
+}
+
+impl PartialEq for Memo {
+    fn eq(&self, rhs: &Memo) -> bool {
+        &self.0[..] == &rhs.0[..]
+    }
+}
+
+impl Memo {
+    /// Returns a `Memo` containing the given slice, appending with zero bytes if
+    /// necessary, or `None` if the slice is too long. If the slice is empty,
+    /// `Memo::default` is returned.
+    pub fn from_bytes(memo: &[u8]) -> Option<Memo> {
+        if memo.is_empty() {
+            Some(Memo::default())
+        } else if memo.len() <= 512 {
+            let mut data = [0; 512];
+            data[0..memo.len()].copy_from_slice(memo);
+            Some(Memo(data))
+        } else {
+            // memo is too long
+            None
+        }
+    }
+
+    /// Returns a `Memo` containing the given string, or `None` if the string is too long.
+    pub fn from_str(memo: &str) -> Option<Memo> {
+        Memo::from_bytes(memo.as_bytes())
+    }
+
+    /// Returns the underlying bytes of the `Memo`.
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.0[..]
+    }
+
+    /// Returns:
+    /// - `None` if the memo is not text
+    /// - `Some(Ok(memo))` if the memo contains a valid UTF-8 string
+    /// - `Some(Err(e))` if the memo contains invalid UTF-8
+    pub fn to_utf8(&self) -> Option<Result<String, str::Utf8Error>> {
+        // Check if it is a text or binary memo
+        if self.0[0] < 0xF5 {
+            // Drop trailing zeroes
```

Comment:
`self.0` is a `[u8; 512]`, not a `str`. We only get a `str` on the next line (because `str` is required to be valid UTF-8 in Rust).

Reply:
Right, but it should be possible to do something like (untested):
```
let s = str::from_utf8(self.0)?;
Some(s.trim_end_matches(char::from(0)));
```
(It's not doing precisely the same thing, but it is equivalent because the trailing NUL characters do not affect whether the string is valid UTF-8.)